doc/http-api.md

```diff
@@ -0,0 +1,112 @@
+# Nitriding's HTTP API
+
+## External endpoints, reachable to the Internet
+
+* `GET /enclave` Returns an index page explaining that this code runs
+  inside an enclave.  
+  The enclave responds with status code `200 OK`.
+
+* `GET /enclave/attestation?nonce={nonce}` Returns an attestation document
+  containing the given nonce.  
+  `nonce` must be a 20-byte nonce encoded in 40 hexadecimal digits.
+  The attestation document is encoded using Base64.
+  If all goes well, the enclave responds with status code `200 OK`.
+
+* `GET /enclave/config` Returns nitriding's configuration.  
+  The enclave responds with status code `200 OK`.
+
+* `GET /enclave/debug` If enabled, returns profiling information.  
+  If nitriding is invoked with the `-debug` command line flag,
+  it exposes this endpoint to make profiling information available.
+  If all goes well, the enclave responds with status code `200 OK`.
+
+## External endpoints, reachable to other enclaves
+
+* `GET /enclave/sync?nonce={nonce}` Exposed by workers, the leader talks to this endpoint to initiate key synchronization.  
+  `nonce` must be a 20-byte nonce encoded in 40 hexadecimal digits.
+  If all goes well, the worker responds with status code `200 OK` and the following JSON-formatted body:
+  ```
+  {
+    "document": "{Base64-encoded attestation document}",
+  }
+  ```
+
+* `POST /enclave/sync` Exposed by workers, the leader talks to this endpoint to
+  complete key synchronization.  
+
+  The request must contain the following JSON-formatted body:
+  ```
+  {
+    "document": "{Base64-encoded attestation document}",
+    "encrypted_keys": "{Base64-encoded, encrypted enclave keys}",
+  }
+  ```
+  If all goes well, the worker responds with status code `200 OK`.
+
+* `POST /enclave/heartbeat` Exposed by the leader, workers periodically send a heartbeat to this endpoint.  
+  The request must contain the following JSON-formatted body:
+  ```
+  {
+    "hashed_keys": "{hashed_keys}",
+    "worker_hostname": "{worker_hostname}",
+  }
+  ```
+  `worker_hostname` contains the worker's EC2-internal hostname, e.g., `ip-12-34-56-78.us-east-2.compute.internal`.  
+  `hashed_keys` contains the Base64-encoded SHA-256 hash over the worker's enclave key material.
+  If all goes well, the leader responds with status code `200 OK`.
+
+* `GET /enclave/leader?nonce={nonce}` Exposed by all enclaves, this endpoint
```

Comment:
For the FQDN is there an internal DNS service running that's configured to point at the leader node as well?

Reply:
There's no internal DNS service. For now, we assume that the domains for both leader and workers are public. That may change though, depending on how the Kubernetes tests are going to go.